### PR TITLE
fix: probe stdio MCP servers via .cmd/.exe fallback on Windows

### PR DESCRIPTION
**Pull request by mrmawk**
## What Changed

  `probeCommandServer` in `scripts/hooks/mcp-health-check.js` now retries the configured stdio command with platform extension candidates on Windows when the bare command name has no extension and is not a path. The probe walks `[command, command.cmd, command.exe, command.bat]` in order and only falls
  through on `ENOENT` / `EINVAL`. `.cmd` / `.bat` candidates are dispatched with `shell: true` so the cmd interpreter (and `PATH` lookup) runs them. Path-like commands (absolute / relative) keep single-candidate ENOENT semantics, so the existing `/ENOENT|spawn/i` failure-reason contract in
  `tests/hooks/mcp-health-check.test.js` is preserved.

  A new Windows-only test `windows: probes bare PATH commands via .cmd fallback` exercises the fallback by writing a `.cmd` shim into a temp dir, prepending it to `PATH`, and asserting the bare command is marked `healthy`. Non-Windows runs print a `- skipped` line and do not touch the pass/fail counts.

  ## Why This Change

  On Windows, `child_process.spawn('npx', …)` raises `ENOENT` because npm-style binaries install as `npx.cmd` and `spawn` does not auto-resolve `.cmd` / `.bat` shims. Worse, since the CVE-2024-27980 mitigation in Node 18.20+ / 20.12+, spawning a `.cmd` target without `shell: true` raises `spawn EINVAL`.
  The combination meant the MCPHealthCheck hook flagged every `npx`-launched stdio MCP (Playwright, Memory, GitHub, Context7, Sequential Thinking, …) as unhealthy on first probe and started blocking their tool calls (`[MCPHealthCheck] <server> is unavailable (spawn npx ENOENT). Blocking <tool> so Claude
  can fall back to non-MCP tools.`), even though Claude Code itself had connected the servers fine.

  This change makes the preflight probe match how the real MCP client launches stdio servers on Windows, eliminating the false-positive block.

  ## Testing Done

  - [x] Manual testing completed
  - [x] Automated tests pass locally (`node tests/run-all.js`)
  - [x] Edge cases considered and tested

  Local results on Windows 11 / Node 24.15.0:
  - `node tests/hooks/mcp-health-check.test.js`: 21 / 21 passed (including new Windows fallback case).
  - `node tests/run-all.js`: 2171 / 2174 passed. The 3 failing tests are pre-existing on `upstream/main` in `scripts/ecc-dashboard.test.js` (`build_terminal_launch …`, `maximize_window …`) and unrelated to this PR — verified by stashing the patch and re-running.
  - `npx eslint scripts/hooks/mcp-health-check.js tests/hooks/mcp-health-check.test.js`: clean.
  - `node scripts/ci/validate-hooks.js`: `Validated 26 hook matchers`.
  - End-to-end: cleared `~/.claude/mcp-health-cache.json`, re-invoked `mcp__playwright__browser_navigate https://example.com`, hook now writes `status: healthy` and the tool call succeeds (previously blocked with `spawn npx ENOENT`).

  ## Type of Change

  - [x] `fix:` Bug fix
  - [ ] `feat:` New feature
  - [ ] `refactor:` Code refactoring
  - [ ] `docs:` Documentation
  - [ ] `test:` Tests
  - [ ] `chore:` Maintenance/tooling
  - [ ] `ci:` CI/CD changes

  ## Security & Quality Checklist

  - [x] No secrets or API keys committed (ghp_, sk-, AKIA, xoxb, xoxp patterns checked)
  - [x] JSON files validate cleanly
  - [x] Shell scripts pass shellcheck (if applicable) — no shell scripts touched
  - [x] Pre-commit hooks pass locally (if configured)
  - [x] No sensitive data exposed in logs or output
  - [x] Follows conventional commits format

  ## Documentation

  - [x] Updated relevant documentation — inline rationale added in code comments; no user-facing doc changed since the hook is internal
  - [x] Added comments for complex logic
  - [x] README updated (if needed) — n/a

---
## Summary by cubic
Fix Windows stdio MCP health checks by probing bare PATH commands with .cmd/.exe/.bat fallbacks. Hardens the probe to avoid EINVAL errors and orphaned processes, preventing false failures (e.g., `npx`) that blocked MCP tools.

- **Bug Fixes**
  - Retry `[command, command.cmd, command.exe, command.bat]` for bare PATH commands; fall back on ENOENT/EINVAL.
  - Use `shell: true` for `.cmd`/`.bat` and kill the process tree with `taskkill /T` on timeout to prevent leaks.
  - Keep path-like commands single-attempt; add a type guard to surface misconfig errors cleanly.

<sup>Written for commit e8f968ea573f75bacbad16915d9131c00db27487. Summary will update on new commits.</sup>

## Summary by CodeRabbit

* **Bug Fixes**
  * More reliable command resolution on Windows with executable-extension fallbacks and per-attempt timeouts.
  * Improved probe termination/cleanup on Windows to avoid orphaned subprocesses and ensure accurate health reporting.

* **Tests**
  * Added a Windows-only test verifying fallback behavior for bare path commands.

### CODE REVIEW

File: scripts/hooks/mcp-health-check.js [hunk 1]
```diff
@@ -306,96 +306,152 @@ function probeCommandServer(serverName, config) {
...
+            // ignore
+          }
+        }, 200).unref?.();
+
+        attemptFinish({
+          ok: true,
+          statusCode: null,
+          reason: `${serverName} accepted a new stdio process`
+        });
+      }, timeoutMs);
```
[the remark this answers is not in the document]
Comment (mrmawk) [reply scripts/hooks/mcp-health-check.js:471]: @greptile-apps thanks for flagging the per-attempt timeout reset. Worth a closer look:

  The earlier candidates only reach `setTimeout(timeoutMs)` if `spawn` succeeds and the process stays alive. In the Windows fallback path, the only outcomes that hand off to the next attempt are `ENOENT` / `EINVAL`, both of which surface synchronously (sync throw from `spawn` or async via
  `child.on('error')` within microseconds — well under 1 ms in practice). The timer is cleared inside `handleEnoent()` before the next attempt arms its own.

  Concretely, the wall-clock cost of the candidate walk is roughly:

  - attempt 0 (`<bare>`) — fast ENOENT, ~ε
  - attempt 1 (`.cmd`) — successful probe → up to `timeoutMs`
  - attempts 2/3 (`.exe`/`.bat`) — only reached if `.cmd` also failed; in that case earlier failures were also ENOENT-fast

  So the realistic upper bound is `≈ timeoutMs + few×ε`, not `4×timeoutMs`. A spawned-but-running candidate would already be the answer (we treat that as healthy on timeout fire).

  I'd rather keep the per-attempt timer than introduce a shared deadline since the existing design relies on `timeoutMs` being the "one process accepted a stdio handshake" budget, not the wall-clock budget for the whole probe. Happy to revisit if you have a repro where the candidate walk actually exceeds
   a single timeoutMs window.

**Comment by mrmawk** [comment]
Thanks @coderabbitai @cubic-dev-ai — addressed in e8f968e:

  - **cubic**: added `commandIsString` type guard before `path.extname` so non-string `config.command` surfaces through `finish()` instead of throwing.
  - **coderabbit (process-tree)**: timer-fired teardown now uses `taskkill /T /F` on Windows when `useShell` is set, walking the cmd subtree and reaping `.cmd` grandchildren. SIGKILL safety net retained.
  - **coderabbit (test cleanup)**: same fix self-cleans the `winfallback.cmd` test shim.
  - **coderabbit (docstring coverage)**: not addressed — repo's existing hook scripts do not use docstrings; happy to revisit if maintainer wants this enforced.

File: scripts/hooks/mcp-health-check.js [hunk 2]
```diff
@@ -306,96 +306,176 @@ function probeCommandServer(serverName, config) {
...
+      path.isAbsolute(command)
+      || command.includes('/')
+      || command.includes('\\')
+    );
+    const candidates = process.platform === 'win32'
+      && commandIsString
+      && !path.extname(command)
+      && !isPathLike
+        ? [command, command + '.cmd', command + '.exe', command + '.bat']
+        : [command];
```
[the remark this answers is not in the document]
Comment (mrmawk) [reply scripts/hooks/mcp-health-check.js:330]: @coderabbitai re: PATHEXT-based ordering — pushing back on this one.

  The candidate order `[bare, .cmd, .exe, .bat]` is intentionally tuned for the real workload this hook serves: stdio MCP servers, where `npx` / `yarn` / `pnpm` dominate. On Windows those tools install **only** as `<name>.cmd` shims (no `.exe` companion), so the existing order resolves on the first
  non-bare candidate with a single fast probe.

  Switching to PATHEXT order (`.COM;.EXE;.BAT;.CMD`) would add a synchronous ENOENT for `.exe` ahead of every npm-style shim — measurable extra latency on every uncached MCP probe, in service of a precedence rule that doesn't apply to any of the commands MCP actually launches. The case it improves (a
  binary that legitimately ships both `foo.exe` and `foo.cmd` and where the `.exe` is preferred) hasn't surfaced as an MCP server pattern in the wild that I'm aware of.

  Reading PATHEXT also adds non-trivial parsing surface (delimiter handling, empty-value fallback, leading-dot variants, case folding, interaction with `path.extname`) that I'd rather not introduce inside a focused bug fix. If a real conflict shows up later, refactoring this into a PATHEXT-aware resolver
   is a clean follow-up — happy to defer to the maintainer's call here.

**Comment by affaan-m** [comment]
Thanks for the Windows MCP stdio probe fallback work. I ported the useful .cmd/.exe/.bat fallback behavior and test coverage onto current `main` in #1719, alongside the existing #1456 path-safety work, so I’m closing this as superseded.
